### ISSUE TITLE
Invalid Stack Pointer in go 1.5beta3

### ISSUE THREAD

**Issue by jonlawlor** [golang/go#11941]
An issue popped up while testing gonum/stat in 1.5beta3: 
https://travis-ci.org/gonum/stat/jobs/73395998
runtime: bad pointer in frame github.com/gonum/matrix/mat64.(*SymDense).CopySym at 0xc820024370: 0x40
fatal error: invalid stack pointer

and also in gonum/matrix:
https://travis-ci.org/gonum/matrix/jobs/73396137
runtime: bad pointer in frame github.com/gonum/matrix/mat64.(*Dense).Clone at 0xc8200432a0: 0x4
fatal error: invalid stack pointer

and also in optimize:
https://travis-ci.org/gonum/optimize/jobs/73396336
runtime: bad pointer in frame github.com/gonum/matrix/mat64.Inner at 0xc820046418: 0x2
fatal error: invalid stack pointer

The same tests pass without panic in 1.4.2 and 1.3.3.

**Comment by ianlancetaylor** [comment]
Those do look like bad stack pointers.

CC @rsc @RLH @aclements

**Comment by dajoo75** [comment]
I get this too intermittently when calling "xml.NewDecoder(f).Decode(r)" e.g.:
  runtime: bad pointer in frame encoding/xml.(*Decoder).Token at 0xc08208d0a8: 0x7
  fatal error: invalid stack pointer

**Comment by aclements** [comment]
I can reproduce this locally with:

```
$ go version
go version devel +e95bc5f Wed Jul 29 18:56:11 2015 +0000 linux/amd64
$ go get -t github.com/gonum/stat/distmv
$ go test -v github.com/gonum/stat/distmv
```

Note that this started in 1.5beta3 because we set the invalidptr GODEBUG to 1 by default (which is what the documentation always claimed it was). I suspect you'd be able to reproduce this on earlier Go 1.5 betas by setting GODEBUG=invalidptr=1 explicitly.

**Comment by btracey** [comment]
The failures seem to be happening in different parts of mat64. Two comments that may help debugging

1) It is not known at compile time if mat64 will use cgo. The mat64 functions call into the blas64 package (github.com/gonum/blas/blas64). The user can "Use" either a native go BLAS implementation or a c-based BLAS implementation. By default, the Go installation is used, and the steps by @aclements  above should be using pure-go code. Possibly though there's some interaction with the possibility of cgo.

2) We have one unusual construct in the mat64 package. As part of the API, it is fine for a user to supply the same concrete value for the receiver and input arguments (for example, a.Mul(a,b), where a is a *mat64.Dense). For most of the methods, the underlying matrix data cannot be shared, so if the pointers are the same, we allocate new storage, and then at the conclusion of the method copy the data over. "at the conclusion" is implemented using a defer statement, that only sometimes executes.
 This particular code can be found in the "isolatedWorkspace" function https://github.com/gonum/matrix/blob/master/mat64/dense.go#L129 . You can see this in use in the add method https://github.com/gonum/matrix/blob/master/mat64/dense_arithmetic.go#L124.  
There have been some changes to the code for closures in 1.5, so possibly one of these introduced a bug.

**Comment by dajoo75** [comment]
Here's a program that recreates the problem on my machine. I'm only using the standard library and the error occurs when calling "xml.NewDecoder(f).Decode(r)" (line 36 in the snippet). There's a comment at the end of the snippet with an example stack trace and go version + go env for my machine):

http://pastebin.com/XHUxVZDH

**Comment by aclements** [comment]
Thanks for the pointers.

Before I go trawling through the mat64 code, are there any places where you might intentionally store a non-pointer in a pointer-typed variable (using unsafe)? I ask because cgo code tends to be susceptible to such things.

**Comment by aclements** [comment]
(Pun not intended. :)

**Comment by btracey** [comment]
We don't use unsafe in the pure-Go code (which again, the code you ran should be pure go). In cgo we do use unsafe, but only in the explicit calls to c functions https://github.com/gonum/blas/blob/master/cgo/blas.go.

We do, however, use asm in some of the routines. Does the failure still happen if you run with -tags=noasm?

**Comment by aclements** [comment]
In all three failures you posted, the bad pointer comes from a frame that's executing "switch b := a.(type)". In two a is a Matrix and in the other it's a Symmetric. The exact same is true of @dajoo75's example, except it's in encoding/xml. It's possible we're generating a bad stack map for this.

**Comment by aclements** [comment]
> We do, however, use asm in some of the routines. Does the failure still happen if you run with -tags=noasm?

noasm doesn't affect it.

**Comment by aclements** [comment]
Building with -gcflags '-S -live=2' reveals that we're passing &b as the "r" argument of runtime.assertI2I2, but we never zeroed b. So, b contains whatever junk was on the stack at that point, but it's considered a live pointer, and the stack scan rightfully complains.

Specifically, we have the sequence
```
60376 (/tmp/11941/src/github.com/gonum/matrix/mat64/symmetric.go:123)	LEAQ	"".b-216(SP), BX
	avarinit=b
60377 (/tmp/11941/src/github.com/gonum/matrix/mat64/symmetric.go:123)	MOVQ	BX, 24(SP)
60692 (/tmp/11941/src/github.com/gonum/matrix/mat64/symmetric.go:123)	PCDATA	$0, $2
60378 (/tmp/11941/src/github.com/gonum/matrix/mat64/symmetric.go:123)	CALL	runtime.assertI2I2(SB)
	live=s,a,autotmp_2790,b
```

but this is the first mention of 216(SP) in this function.

**Comment by aclements** [comment]
Turns out @rsc fixed this last night with c4092ac, but that was after beta 3 was cut. Once the next beta rolls out, you should be good.
